File: test-suite/golang/security/header_injection_clean.go
package security

import (
	"mime"
	"net/http"
	"strings"
)

type cleanHeaderContext interface {
	Query(string) string
	Header(string, string)
}

func safeHeaderValue(raw string) string {
	return strings.NewReplacer("\r", "", "\n", "").Replace(raw)
}

func queryValueInSafeHeader(w http.ResponseWriter, r *http.Request) {
	displayName := safeHeaderValue(r.URL.Query().Get("name"))
	w.Header().Set("X-Display-Name", displayName)
}

func safeContentDisposition(w http.ResponseWriter, r *http.Request) {
	filename := r.FormValue("filename")
	disposition := mime.FormatMediaType("attachment", map[string]string{"filename": filename})
	w.Header().Set("Content-Disposition", disposition)
}

func guardedRequestHeader(w http.ResponseWriter, r *http.Request) {
	traceID := r.Header.Get("X-Trace-ID")
	if strings.ContainsAny(traceID, "\r\n") {
		http.Error(w, "invalid trace header", http.StatusBadRequest)
		return
	}
	w.Header().Add("X-Upstream-Trace", traceID)
}

func frameworkHeaderWithSafeHelper(c cleanHeaderContext) {
	c.Header("X-Return-Reason", safeHeaderValue(c.Query("reason")))
}

File: test-suite/csharp/security/HeaderInjectionClean.cs
using System;
using System.Net;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Primitives;

public sealed class HeaderInjectionClean : Controller
{
    private static string SafeHeaderValue(string raw)
    {
        return raw.Replace("\r", string.Empty).Replace("\n", string.Empty);
    }

    public void QueryValueInSafeHeader(HttpContext context)
    {
        var displayName = SafeHeaderValue(context.Request.Query["name"]!);
        context.Response.Headers["X-Display-Name"] = displayName;
    }

    public void EncodedFilenameDisposition(HttpContext context)
    {
        var filename = WebUtility.UrlEncode(context.Request.Form["filename"]!);
        context.Response.Headers.Append("Content-Disposition", "attachment; filename=" + filename);
    }

    public IActionResult GuardedRequestHeader(HttpContext context)
    {
        context.Request.Headers.TryGetValue("X-Trace-ID", out StringValues traceId);
        var trace = traceId.ToString();
        if (trace.Contains("\r", StringComparison.Ordinal) ||
            trace.Contains("\n", StringComparison.Ordinal))
        {
            return BadRequest();
        }

        context.Response.Headers.Add("X-Upstream-Trace", trace);
        return Ok();
    }

    public void AnnotatedQueryParameter([FromQuery] string reason)
    {
        Response.Headers.Set("X-Return-Reason", SafeHeaderValue(reason));
    }
}
